libs/langchain/langchain/retrievers/parent_document_retriever_v2.py

```diff
@@ -0,0 +1,116 @@
+import uuid
+from typing import List, Optional, Sequence
+
+from langchain.retrievers import MultiVectorRetriever
+from langchain.schema.document import BaseDocumentTransformer, Document
+
+
+class ParentDocumentRetriever(MultiVectorRetriever):
+    """Retrieve small chunks then retrieve their parent documents.
+
+    When transform documents for retrieval, there are often conflicting desires:
+
+    1. You may want to have small documents, so that their embeddings can most
+        accurately reflect their meaning. If too long, then the embeddings can
+        lose meaning.
+    2. You want to have long enough documents that the context of each chunk is
+        retained.
+
+    The NewParentDocumentRetriever strikes that balance by transforming and storing
+    small chunks of data. During retrieval, it first fetches the small chunks
+    but then looks up the parent ids for those chunks and returns those larger
+    documents.
+
+    Note that "parent document" refers to the document that a small chunk
+    originated from. This can either be the whole raw document OR a larger
+    chunk.
+
+    Examples:
+
+        .. code-block:: python
+
+            # Imports
+            from langchain.vectorstores import Chroma
+            from langchain.embeddings import OpenAIEmbeddings
+            from langchain.text_splitter import RecursiveCharacterTextSplitter
+            from langchain.storage import InMemoryStore
+
+            # This text splitter is used to create the parent documents
+            parent_transformer = RecursiveCharacterTextSplitter(chunk_size=2000)
+            # This text splitter is used to create the child documents
+            # It should create documents smaller than the parent
+            child_transformer = RecursiveCharacterTextSplitter(chunk_size=400)
+            # The vectorstore to use to index the child chunks
+            vectorstore = Chroma(embedding_function=OpenAIEmbeddings())
+            # The storage layer for the parent documents
+            store = InMemoryStore()
+
+            # Initialize the retriever
+            retriever = ParentDocumentRetriever(
+                vectorstore=vectorstore,
+                docstore=store,
+                child_transformer=child_transformer,
+                parent_transformer=parent_transformer,
+            )
+    """
+
+    child_transformer: BaseDocumentTransformer
```

Comment:
@pprados let's change this on the original parent document retriever.

If there is a way to change names using pydantic features without making breaking changes that's great. If not, that's fine -- better have a slight unintuitive parameter name than a separate class with essentially identical functionality

Reply:
@eyurtsev, I am working on a better implementation.

A question about the usage of parser.
When I use a parser in a prompt, now, the method predict_and_parse() throw a warning
`The predict_and_parse method is deprecated, instead pass an output parser directly to LLMChain.`.

Does this mean that it is no longer possible to have a parser in a prompt?
However, this is a good idea for complex algorithms requiring several prompts, such as map_reduce, refine, etc.
I don't know how it's possible to pass an output parser directly to LLMChain. To do this, I must always receive an LLMChain and not a prompt as a parameter.

The `map_reduce_chain` receive a `question_prompt`, ` combine_prompt` and `collapse_prompt`. It will be deprecated ?

Reply:
> I am working on a better implementation.

Sounds good -- I would just suggest to discuss implementation / design prior to putting in the effort to make sure that the new code gets merged in

---

1) LLMChain is a legacy object. It's not a good idea to build new functionality on top of it. Use LCEL runnables instead.

2) LLMChain accepts a parser: https://github.com/langchain-ai/langchain/blob/68599d98c20b1e3fbdabaef1b1fbe54cd06b98a4/libs/langchain/langchain/chains/llm.py#L54C1-L54C1. (But again we don't want to be using LLMChain in new code)

3) The map_reduce_chain receive a question_prompt, combine_prompt and collapse_prompt -- no plans to deprecate them as users are using them. We may at some point provide new implementations with LCEL